### PR TITLE
[+] MongoDB HasMany - Allow documents embedded to an array to be editable

### PR DESCRIPTION
**Pull request by ArnaudValensi**
## Pull Request checklist:

- [x] Write an explicit title for the Pull Request
- [x] Write changes made in the CHANGELOG.md
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

### CODE REVIEW

**Comment by ArnaudValensi** [comment]
[(3) Embedded HasMany - Friendo wants to edit documents in array of embedded documents](https://trello.com/c/mh9laEoU/2265-3-embedded-hasmany-friendo-wants-to-edit-documents-in-array-of-embedded-documents)

**Comment by ArnaudValensi** [comment]
[(3) Embedded HasMany - Friendo wants to edit documents in array of embedded documents](https://trello.com/c/mh9laEoU/2265-3-embedded-hasmany-friendo-wants-to-edit-documents-in-array-of-embedded-documents)

File: services/embedded-resource-updater.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      }
+
+      var data = record[subDocumentName];
+
+      var update = Object.keys(data).reduce(function (acc, value) {
+        acc.$set[subDocumentName + '.' + recordIndex + '.' + value] = data[value];
+        return acc;
+      }, { $set: {} });
+
+      model.update({ _id: id }, update, function (err) {
```
Comment (arnaudbesnier) [inline services/embedded-resource-updater.js]: I think you should use `findByIdAndUpdate` to abstract and remove `_id:`.

File: services/embedded-resource-updater.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        if (err) { return reject(err); }
```
Comment (arnaudbesnier) [inline services/embedded-resource-updater.js]: `err` => `error`

File: services/embedded-resource-updater.js (hunk 1 above)
Comment (arnaudbesnier) [reply services/embedded-resource-updater.js]: `return model.update({ ... });`

File: services/embedded-resource-updater.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    return new P(function (resolve, reject) {
+      var recordId = params.recordId;
+      var recordIndex = parseInt(params.recordIndex, 10);
+
+      delete record._id;
+
+      var id = mongoose.Types.ObjectId(recordId);
+      var subDocumentName = Object.keys(record)[0];
+
+      if (!subDocumentName || !subDocumentName.length) {
```
Comment (arnaudbesnier) [inline services/embedded-resource-updater.js]: What is `!subDocumentName.length` for?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,8 @@
 # Change Log
 
 ## [Unreleased]
+### Added
+- MongoDB HasMany - Allow documents embedded to an array to be editable.
```
Comment (arnaudbesnier) [inline CHANGELOG.md]: Users can edit a document in an array of embedded documents?

File: services/embedded-resource-updater.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+var P = require('bluebird');
+var mongoose = require('mongoose');
+var createError = require('http-errors');
+
+function EmbeddedResourceUpdater(model, params, record) {
+  this.perform = function () {
+    return new P(function (resolve, reject) {
```
Comment (arnaudbesnier) [inline services/embedded-resource-updater.js]: Is the promise really necessary?

Everything looks synchronous, except the `model.update`. So you just have to return it `return model.update(...)`.

File: services/embedded-resource-updater.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      var data = record[subDocumentName];
+
+      var update = Object.keys(data).reduce(function (acc, value) {
+        acc.$set[subDocumentName + '.' + recordIndex + '.' + value] = data[value];
+        return acc;
+      }, { $set: {} });
+
+      model.update({ _id: id }, update, function (err) {
+        if (err) { return reject(err); }
+        console.log('ok');
```
[the remark this answers is not in the document]
Comment (ArnaudValensi) [reply services/embedded-resource-updater.js]: Done

File: services/embedded-resource-updater.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+var P = require('bluebird');
+var mongoose = require('mongoose');
+var createError = require('http-errors');
+
+function EmbeddedResourceUpdater(model, params, record) {
```
[the remark this answers is not in the document]
Comment (ArnaudValensi) [reply services/embedded-resource-updater.js]: Done

File: services/embedded-resource-updater.js (hunk 1 above)
Comment (ArnaudValensi) [reply services/embedded-resource-updater.js]: Done

File: services/embedded-resource-updater.js (hunk 2 above)
Comment (ArnaudValensi) [reply services/embedded-resource-updater.js]: Done

File: services/embedded-resource-updater.js (hunk 3 above)
Comment (ArnaudValensi) [reply services/embedded-resource-updater.js]: Done

File: CHANGELOG.md (hunk 4 above)
Comment (ArnaudValensi) [reply CHANGELOG.md]: Done

File: services/embedded-resource-updater.js (hunk 5 above)
Comment (ArnaudValensi) [reply services/embedded-resource-updater.js]: Done

**Pull request merged**